Create symlink support

jelly opened the pull request:
Allow the creation of absolute or relative symlinks from the target. The symlink input field can be a filename, an absolute or relative path and are either symlinked as relative or absolute depending on the selected option.

The symlink permissions are not configurable and by default based on the owner of the directory so creating a symlink as administrator in the users home directory creates it as `admin:admin`.

Closes: #391

---

# Files: symlink creation support

Symlinks can now be created for files and directories by right clicking and clicking `Create link` menu entry. Only relative or absolute symlinks are supported.

![image](https://github.com/user-attachments/assets/f4250584-3f28-455d-b8f8-502cd9e94e4e)

jelly commented:
See comments in https://github.com/cockpit-project/cockpit-files/pull/959

martinpitt reviewed (changes requested):
Thanks! My review was at https://github.com/cockpit-project/cockpit-files/pull/959#pullrequestreview-2680945800 , I'm sorry for the bad timing. I'll do the next round of review here of course.

jelly commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+        cockpit.assert(cwdInfo !== null);
+
+        const cmd = ["ln", "--symbolic", "--no-target-directory"];
+        const target = `${path}${selected.name}`;
```
Not applicable, not nodejs code :)

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+        if (symlinkName.startsWith('/') || symlinkName.startsWith('..')) {
+            try {
+                folder_info = await fsinfo(dirname(symlinkPath), ["user", "group", "mode"], { superuser: "try" });
+            } catch (exc) {
+                const err = exc as BasicError;
+                setErrorMessage(err.toString());
+                console.warn(`fsinfo failed for ${symlinkPath}`, exc);
+                return;
```
These 5 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+          variant={ModalVariant.medium}
+          isOpen
+          className="file-symlink-modal"
+          onClose={() => dialogResult.resolve()}
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+                  <Button
+                    variant="primary"
+                    onClick={() => {
+                        const filename_invalid = checkLinkName(symlinkName, cwdInfo?.entries || {});
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+            />}
+            <Form
+              isHorizontal
+              onSubmit={e => {
+                  e.preventDefault();
+                  createLink();
+                  return false;
```
These 4 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+                      validated={symlinkError ? "error" : "default"}
+                      value={symlinkName}
+                      onChange={(_, val) => {
+                          const filename_invalid = checkLinkName(val, cwdInfo?.entries || {});
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,243 @@
+                      name="relative"
+                      label={_("Relative")}
+                      isChecked={mode === "relative"}
+                      onChange={() => setMode("relative")}
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+        if (symlinkName.startsWith('/') || symlinkName.startsWith('..')) {
+            try {
+                folder_info = await fsinfo(dirname(symlinkPath), ["user", "group", "mode"], { superuser: "try" });
+            } catch (exc) {
+                const err = exc as BasicError;
+                setErrorMessage(err.toString());
+                console.warn(`fsinfo failed for ${symlinkPath}`, exc);
+                return;
```
These 5 added lines are not executed by any test.

cockpituous replied:
These 5 added lines are not executed by any test.

cockpituous replied:
These 5 added lines are not executed by any test.

cockpituous replied:
These 5 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+          variant={ModalVariant.medium}
+          isOpen
+          className="file-symlink-modal"
+          onClose={() => dialogResult.resolve()}
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                  <Button
+                    variant="primary"
+                    onClick={() => {
+                        const filename_invalid = checkLinkName(symlinkName, cwdInfo?.entries || {});
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+            />}
+            <Form
+              isHorizontal
+              onSubmit={e => {
+                  e.preventDefault();
+                  createLink();
+                  return false;
```
These 4 added lines are not executed by any test.

cockpituous replied:
These 4 added lines are not executed by any test.

cockpituous replied:
These 4 added lines are not executed by any test.

cockpituous replied:
These 4 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                      validated={symlinkError ? "error" : "default"}
+                      value={symlinkName}
+                      onChange={(_, val) => {
+                          const filename_invalid = checkLinkName(val, cwdInfo?.entries || {});
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                      name="relative"
+                      label={_("Relative")}
+                      isChecked={mode === "relative"}
+                      onChange={() => setMode("relative")}
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

jelly commented:
@martinpitt re-pushed and changed:

* remove the "link to" highlighting, it is indeed broken and triggers every time
* simplified `realpath` but only when a non-absolute path is given
* dropped the useless button
* switched to `--no-target-directory` so we don't get an weird behaviour such as specifying /opt which creates /opt/bar instead of trying to create the symlink /opt

martinpitt commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+            await cockpit.spawn(cmd, {
+                err: "message",
+                directory: path,
+                ...owner && { superuser: "require" }
```
I had to stare at this for a bit. It doesn't mean `(...owner) && {..}` but `...(owner && {..})`.

jelly replied:
that makes hmm no sense, actually this whole thing is interesting right?

```
        if (superuser.allowed) {
            owner = [...get_owner_candidates(folder_info)][0];
        }
```

So basically only adding superuser: "require" when it is allowed, so we could change this to "try" imo.

martinpitt replied:
I meant that the code is right, it's just not very obvious with that spacing and the lack of parentheses. If `owner` is `null`, then this resolves to `{...false} === {}`, and otherwise to `{ superuser: "require" }`

martinpitt replied:
I.e. this was just thinking aloud, you don't have to change it and this thread can just be resolved.

jelly replied:
But ...(owner && {}) is but I am not sure if it is more readable?

jelly replied:
Because this is a common pattern:

```
        setAlerts(prevAlerts => [
            ...prevAlerts, {
                title,
                variant,
                key,
                ...detail && { detail },
                ...actionLinks && { actionLinks },
            }
```

martinpitt replied:
's all good :grin:

martinpitt commented on src/menu.tsx:
```diff
@@ -151,6 +152,15 @@ export function get_menu_items(
+        if (item.type === 'reg' || item.type === "dir")
+            menuItems.push(
+                {
+                    id: "create-symlink",
```
This is now the first or second entry in the menu. It feels more like a fringe feature to me, how about we move that to the bottom?

jelly replied:
Sure, I've asked Garrett for the placement but never got a reply.

jelly replied:
Like this ?
![image](https://github.com/user-attachments/assets/131b0396-3c0c-4129-941d-e5f66a53268a)

martinpitt commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                          setErrorMessage(undefined);
+                          setSymlinkName(val);
+                      }}
+                      id="symlink-input" autoFocus // eslint-disable-line jsx-a11y/no-autofocus
```
This doesn't automatically select the "link to filename" text, so it has to be manually deleted now. Can't we do that as [`placeholder=`](https://v5-archive.patternfly.org/components/forms/text-input) instead? (This is a design question, and I don't claim authority here -- but we do use this a lot in cockpit and c-files).

jelly replied:
A placeholder is not auto selected right? I think I will make a follow up issue for this.

martinpitt replied:
Yes, the `autoFocus` should stay of course, it was just a convenient place to attach the comment to. You can't "select" the placeholder itself, it's not an actionable text.

jelly replied:
We can't do a place holder as the whole idea was that the default symlink target would be `Link to $0`. I will create a follow up issue as the idea is, to have it pre-select "link to" and allow you to quickly remove that once.

martinpitt replied:
But this isn't a sensible default -- nobody actually wants a `link to myfile`. It's an explanation, not a reasonable default. And there is no reasonable default, hence it should be empty.

martinpitt replied:
Compare this to "create new file" where we also don't pre-fill the file name.

martinpitt reviewed (changes requested):
Thank you! This makes much more sense and is more robust. We can fine-tune this based on user feedback/SPUR in follow-ups.

I cleaned up my [original UI review](https://github.com/cockpit-project/cockpit-files/pull/959#pullrequestreview-2680945800), there is nothing relevant left any more.

I have two more UI suggestions, but those aren't strong opinions. If you are like "I need to get this out of my eyes", I'm fine with converting to :+1: .

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+type symlink_mode = "absolute" | "relative";
+
+function checkLinkName(candidate: string, entries: Record<string, FileInfo>) {
+    if (candidate === "") {
+        return _("Name cannot be empty");
+    } else if (candidate in entries) {
+        if (entries[candidate].type === "dir") {
+            return _("Directory with the same name exists");
```
These 5 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+        if (entries[candidate].type === "dir") {
+            return _("Directory with the same name exists");
+        }
+        return _("File exists");
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+            try {
+                symlinkPath = await cockpit.spawn(["realpath", `${path}${symlinkName}`], { superuser: "try" });
+                symlinkPath = symlinkPath.trim();
+            } catch (exc) {
+                const err = exc as BasicError;
+                setErrorMessage(err.toString());
+                console.warn("unable to obtain realpath", exc);
+                return;
```
These 5 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                    superuser: "require"
+                });
+            }
+        } catch (exc) {
+            const err = exc as BasicError;
+            setErrorMessage(err.toString());
+            return;
```
These 4 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                  <Button
+                    variant="primary"
+                    onClick={() => {
+                        const filename_invalid = checkLinkName(symlinkName, cwdInfo?.entries || {});
+                        if (filename_invalid) {
+                            setSymlinkError(filename_invalid);
```
These 3 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                  >
+                      {_("Create link")}
+                  </Button>
+                  <Button variant="link" onClick={() => dialogResult.resolve()}>{_("Cancel")}</Button>
```
This added line is not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+          }
+        >
+            {errorMessage &&
+            <InlineNotification
+              type="danger"
+              text={errorMessage}
+              isInline
```
These 4 added lines are not executed by any test.

cockpituous commented on src/dialogs/symlink.tsx:
```diff
@@ -0,0 +1,244 @@
+                </FormGroup>
+                <FormGroup fieldId="symlink-input" label={_("Symlink name")}>
+                    <TextInput
+                      validated={symlinkError ? "error" : "default"}
```
This added line is not executed by any test.

cockpituous commented on src/menu.tsx:
```diff
@@ -185,6 +186,15 @@ export function get_menu_items(
                     onClick: () => downloadFile(path, item)
                 }
             );
+            if (item.type === 'reg' || item.type === "dir")
```
This added line is not executed by any test.

martinpitt commented:
Anyway, enough of my bitching -- I'd say, please add the new pixel refs and we land that *censored* thing? :grin: We can iterate in the future.

martinpitt reviewed (approved):
"approval through exhaustion" :grin: Thanks!
